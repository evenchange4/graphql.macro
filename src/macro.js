// @flow
import { createMacro } from 'babel-plugin-macros';
import gqlTag from 'graphql-tag';
import serialize from 'babel-literal-to-ast';
import {
  expandImports,
  compileWithFragment,
  resolveImportPath,
} from './utils/index';
// import printAST from 'ast-pretty-print';
// console.log(printAST(referencePath.parentPath))

function graphqlMacro({
  references,
  state: {
    file: {
      opts: { filename },
    },
  },
  babel: { types: t },
}: {
  references: { gql: Array<any>, loader: Array<any> },
  state: { file: { opts: { filename: string } } },
  babel: { types: Object },
}): void {
  const { gql = [], loader = [] } = references;

  // Case 1: import { gql } from 'graphql.macro'.
  gql.forEach(referencePath => {
    const compiled = compileWithFragment(referencePath, t);
    referencePath.parentPath.replaceWith(compiled);
  });

  // Case 2: import { loader } from 'graphql.macro'
  loader.forEach(referencePath => {
    referencePath.parentPath.node.arguments.forEach(({ value }) => {
      const absolutePath = resolveImportPath({ filename, relativePath: value });
      const expanded = expandImports(absolutePath); // Note: #import feature
      referencePath.parentPath.replaceWith(serialize(gqlTag(expanded)));
    });
  });
}

export default createMacro(graphqlMacro);
