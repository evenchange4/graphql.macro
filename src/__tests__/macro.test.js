// @flow

import pluginTester from 'babel-plugin-tester';
import plugin from 'babel-plugin-macros';
import fs from 'fs';
import path from 'path';

expect.addSnapshotSerializer({
  print(val) {
    return val.replace(/..\/macro/, 'graphql.macro');
  },
  test(val) {
    return typeof val === 'string';
  },
});

pluginTester({
  plugin,
  snapshot: true,
  babelOptions: {
    filename: __filename,
  },
  tests: {
    '[gql] without fragment': {
      error: false,
      code: `
        import { gql } from '../macro';
        const query = gql\`
          query User {
            user(id: 5) {
              lastName
              ...UserEntry1
            }
          }
        \`;
      `,
    },
    '[gql] with fragment': {
      error: false,
      code: `
        import { gql } from '../macro';
        const userFragment = gql\`
          fragment UserEntry1 on User {
            firstName
          }
        \`;
        const query = gql\`
          query User {
            user(id: 5) {
              lastName
              ...UserEntry1
            }
          }
          $\{userFragment}
        \`;
      `,
    },
    '[loader] without fragment': {
      error: false,
      code: `
        import { loader } from '../macro';
        const query = loader('./fixtures/query.graphql');
      `,
    },
    '[loader] with fragment': {
      error: false,
      code: `
        import { loader } from '../macro';
        const query = loader('./fixtures/query1.graphql');
      `,
    },
    '[loader] with absolute path': {
      error: false,
      code: `
        import { loader } from '../macro';
        const query = loader('src/__tests__/fixtures/simpleFragment.graphql');
      `,
    },
    '[loader] with absolute path and NODE_PATH': {
      error: false,
      code: `
        import { loader } from '../macro';
        const query = loader('__tests__/fixtures/simpleFragment.graphql');
      `,
      setup: () => {
        process.env.NODE_PATH = 'src/';
      },
    },
    '[loader] with absolute path and jsconfig include property': {
      error: false,
      code: `
        import { loader } from '../macro';
        const query = loader('__tests__/fixtures/simpleFragment.graphql');
      `,
      setup: () => {
        fs.symlinkSync(
          'src/__tests__/fixtures/jsconfig.json',
          path.resolve(fs.realpathSync(process.cwd()), 'jsconfig.json'),
          'file',
        );
      },
      teardown: () => {
        fs.unlinkSync(
          path.resolve(fs.realpathSync(process.cwd()), 'jsconfig.json'),
        );
      },
    },
    '[loader] with nested circular fragments': {
      error: false,
      code: `
        import { loader } from '../macro';
        const query = loader('./fixtures/query3.graphql');
      `,
    },
    '[loader] should work with relative path issue#52': {
      error: false,
      code: `
        import { loader } from '../macro';
        const query = loader('../__tests__/fixtures/query.graphql');
      `,
    },
    // '[loader] multiple operations': {
    //   error: false,
    //   code: `
    //     import { loader } from '../macro';
    //     const { User, Post } = loader('./fixtures/query2.graphql');
    //   `,
    // },
  },
});
