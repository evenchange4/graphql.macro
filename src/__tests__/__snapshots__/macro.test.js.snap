// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros [gql] with fragment: [gql] with fragment 1`] = `

import { gql } from 'graphql.macro';
const userFragment = gql\`
  fragment UserEntry1 on User {
    firstName
  }
\`;
const query = gql\`
  query User {
    user(id: 5) {
      lastName
      ...UserEntry1
    }
  }
  \${userFragment}
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

const userFragment = {
  "kind": "Document",
  "definitions": [{
    "kind": "FragmentDefinition",
    "name": {
      "kind": "Name",
      "value": "UserEntry1"
    },
    "typeCondition": {
      "kind": "NamedType",
      "name": {
        "kind": "Name",
        "value": "User"
      }
    },
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "firstName"
        },
        "arguments": [],
        "directives": []
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 51,
    "source": {
      "body": "\\n  fragment UserEntry1 on User {\\n    firstName\\n  }\\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};
const query = {
  "kind": "Document",
  "definitions": [{
    "kind": "OperationDefinition",
    "operation": "query",
    "name": {
      "kind": "Name",
      "value": "User"
    },
    "variableDefinitions": [],
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "user"
        },
        "arguments": [{
          "kind": "Argument",
          "name": {
            "kind": "Name",
            "value": "id"
          },
          "value": {
            "kind": "IntValue",
            "value": "5"
          }
        }],
        "directives": [],
        "selectionSet": {
          "kind": "SelectionSet",
          "selections": [{
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "lastName"
            },
            "arguments": [],
            "directives": []
          }, {
            "kind": "FragmentSpread",
            "name": {
              "kind": "Name",
              "value": "UserEntry1"
            },
            "directives": []
          }]
        }
      }]
    }
  }].concat(userFragment.definitions),
  "loc": {
    "start": 0,
    "end": 82,
    "source": {
      "body": "\\n  query User {\\n    user(id: 5) {\\n      lastName\\n      ...UserEntry1\\n    }\\n  }\\n  \\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};

`;

exports[`macros [gql] without fragment: [gql] without fragment 1`] = `

import { gql } from 'graphql.macro';
const query = gql\`
  query User {
    user(id: 5) {
      lastName
      ...UserEntry1
    }
  }
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

const query = {
  "kind": "Document",
  "definitions": [{
    "kind": "OperationDefinition",
    "operation": "query",
    "name": {
      "kind": "Name",
      "value": "User"
    },
    "variableDefinitions": [],
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "user"
        },
        "arguments": [{
          "kind": "Argument",
          "name": {
            "kind": "Name",
            "value": "id"
          },
          "value": {
            "kind": "IntValue",
            "value": "5"
          }
        }],
        "directives": [],
        "selectionSet": {
          "kind": "SelectionSet",
          "selections": [{
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "lastName"
            },
            "arguments": [],
            "directives": []
          }, {
            "kind": "FragmentSpread",
            "name": {
              "kind": "Name",
              "value": "UserEntry1"
            },
            "directives": []
          }]
        }
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 79,
    "source": {
      "body": "\\n  query User {\\n    user(id: 5) {\\n      lastName\\n      ...UserEntry1\\n    }\\n  }\\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};

`;

exports[`macros [loader] should work with relative path issue#52: [loader] should work with relative path issue#52 1`] = `

import { loader } from 'graphql.macro';
const query = loader('../__tests__/fixtures/query.graphql');

      ↓ ↓ ↓ ↓ ↓ ↓

const query = {
  "kind": "Document",
  "definitions": [{
    "kind": "OperationDefinition",
    "operation": "query",
    "name": {
      "kind": "Name",
      "value": "User"
    },
    "variableDefinitions": [],
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "user"
        },
        "arguments": [{
          "kind": "Argument",
          "name": {
            "kind": "Name",
            "value": "id"
          },
          "value": {
            "kind": "IntValue",
            "value": "5"
          }
        }],
        "directives": [],
        "selectionSet": {
          "kind": "SelectionSet",
          "selections": [{
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "firstName"
            },
            "arguments": [],
            "directives": []
          }, {
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "lastName"
            },
            "arguments": [],
            "directives": []
          }]
        }
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 62,
    "source": {
      "body": "query User {\\n  user(id: 5) {\\n    firstName\\n    lastName\\n  }\\n}\\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};

`;

exports[`macros [loader] with absolute path and NODE_PATH: [loader] with absolute path and NODE_PATH 1`] = `

import { loader } from 'graphql.macro';
const query = loader('__tests__/fixtures/simpleFragment.graphql');

      ↓ ↓ ↓ ↓ ↓ ↓

const query = {
  "kind": "Document",
  "definitions": [{
    "kind": "FragmentDefinition",
    "name": {
      "kind": "Name",
      "value": "UserEntry1"
    },
    "typeCondition": {
      "kind": "NamedType",
      "name": {
        "kind": "Name",
        "value": "User"
      }
    },
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "firstName"
        },
        "arguments": [],
        "directives": []
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 44,
    "source": {
      "body": "fragment UserEntry1 on User {\\n  firstName\\n}\\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};

`;

exports[`macros [loader] with absolute path and jsconfig include property: [loader] with absolute path and jsconfig include property 1`] = `

import { loader } from 'graphql.macro';
const query = loader('__tests__/fixtures/simpleFragment.graphql');

      ↓ ↓ ↓ ↓ ↓ ↓

const query = {
  "kind": "Document",
  "definitions": [{
    "kind": "FragmentDefinition",
    "name": {
      "kind": "Name",
      "value": "UserEntry1"
    },
    "typeCondition": {
      "kind": "NamedType",
      "name": {
        "kind": "Name",
        "value": "User"
      }
    },
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "firstName"
        },
        "arguments": [],
        "directives": []
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 44,
    "source": {
      "body": "fragment UserEntry1 on User {\\n  firstName\\n}\\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};

`;

exports[`macros [loader] with absolute path: [loader] with absolute path 1`] = `

import { loader } from 'graphql.macro';
const query = loader('src/__tests__/fixtures/simpleFragment.graphql');

      ↓ ↓ ↓ ↓ ↓ ↓

const query = {
  "kind": "Document",
  "definitions": [{
    "kind": "FragmentDefinition",
    "name": {
      "kind": "Name",
      "value": "UserEntry1"
    },
    "typeCondition": {
      "kind": "NamedType",
      "name": {
        "kind": "Name",
        "value": "User"
      }
    },
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "firstName"
        },
        "arguments": [],
        "directives": []
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 44,
    "source": {
      "body": "fragment UserEntry1 on User {\\n  firstName\\n}\\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};

`;

exports[`macros [loader] with fragment: [loader] with fragment 1`] = `

import { loader } from 'graphql.macro';
const query = loader('./fixtures/query1.graphql');

      ↓ ↓ ↓ ↓ ↓ ↓

const query = {
  "kind": "Document",
  "definitions": [{
    "kind": "FragmentDefinition",
    "name": {
      "kind": "Name",
      "value": "UserEntry1"
    },
    "typeCondition": {
      "kind": "NamedType",
      "name": {
        "kind": "Name",
        "value": "User"
      }
    },
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "firstName"
        },
        "arguments": [],
        "directives": []
      }]
    }
  }, {
    "kind": "FragmentDefinition",
    "name": {
      "kind": "Name",
      "value": "UserEntry3"
    },
    "typeCondition": {
      "kind": "NamedType",
      "name": {
        "kind": "Name",
        "value": "User"
      }
    },
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "id"
        },
        "arguments": [],
        "directives": []
      }]
    }
  }, {
    "kind": "FragmentDefinition",
    "name": {
      "kind": "Name",
      "value": "UserEntry2"
    },
    "typeCondition": {
      "kind": "NamedType",
      "name": {
        "kind": "Name",
        "value": "User"
      }
    },
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "lastName"
        },
        "arguments": [],
        "directives": []
      }]
    }
  }, {
    "kind": "OperationDefinition",
    "operation": "query",
    "name": {
      "kind": "Name",
      "value": "User"
    },
    "variableDefinitions": [],
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "user"
        },
        "arguments": [{
          "kind": "Argument",
          "name": {
            "kind": "Name",
            "value": "id"
          },
          "value": {
            "kind": "IntValue",
            "value": "5"
          }
        }],
        "directives": [],
        "selectionSet": {
          "kind": "SelectionSet",
          "selections": [{
            "kind": "FragmentSpread",
            "name": {
              "kind": "Name",
              "value": "UserEntry1"
            },
            "directives": []
          }, {
            "kind": "FragmentSpread",
            "name": {
              "kind": "Name",
              "value": "UserEntry2"
            },
            "directives": []
          }]
        }
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 257,
    "source": {
      "body": "fragment UserEntry1 on User {\\n  firstName\\n}\\n\\nfragment UserEntry3 on User {\\n  id\\n}\\nfragment UserEntry2 on User {\\n  lastName\\n}\\n#import \\"./fragment1.graphql\\"\\n#import './fragment2.graphql'\\n\\nquery User {\\n  user(id: 5) {\\n    ...UserEntry1\\n    ...UserEntry2\\n  }\\n}\\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};

`;

exports[`macros [loader] with nested circular fragments: [loader] with nested circular fragments 1`] = `

import { loader } from 'graphql.macro';
const query = loader('./fixtures/query3.graphql');

      ↓ ↓ ↓ ↓ ↓ ↓

const query = {
  "kind": "Document",
  "definitions": [{
    "kind": "FragmentDefinition",
    "name": {
      "kind": "Name",
      "value": "UserEntry1"
    },
    "typeCondition": {
      "kind": "NamedType",
      "name": {
        "kind": "Name",
        "value": "User"
      }
    },
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "firstName"
        },
        "arguments": [],
        "directives": []
      }]
    }
  }, {
    "kind": "FragmentDefinition",
    "name": {
      "kind": "Name",
      "value": "UserEntry3"
    },
    "typeCondition": {
      "kind": "NamedType",
      "name": {
        "kind": "Name",
        "value": "User"
      }
    },
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "id"
        },
        "arguments": [],
        "directives": []
      }]
    }
  }, {
    "kind": "FragmentDefinition",
    "name": {
      "kind": "Name",
      "value": "UserEntry5"
    },
    "typeCondition": {
      "kind": "NamedType",
      "name": {
        "kind": "Name",
        "value": "User"
      }
    },
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "FragmentSpread",
        "name": {
          "kind": "Name",
          "value": "UserEntry4"
        },
        "directives": []
      }]
    }
  }, {
    "kind": "FragmentDefinition",
    "name": {
      "kind": "Name",
      "value": "UserEntry4"
    },
    "typeCondition": {
      "kind": "NamedType",
      "name": {
        "kind": "Name",
        "value": "User"
      }
    },
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "FragmentSpread",
        "name": {
          "kind": "Name",
          "value": "UserEntry1"
        },
        "directives": []
      }, {
        "kind": "FragmentSpread",
        "name": {
          "kind": "Name",
          "value": "UserEntry5"
        },
        "directives": []
      }]
    }
  }, {
    "kind": "OperationDefinition",
    "operation": "query",
    "name": {
      "kind": "Name",
      "value": "User"
    },
    "variableDefinitions": [],
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "user"
        },
        "arguments": [{
          "kind": "Argument",
          "name": {
            "kind": "Name",
            "value": "id"
          },
          "value": {
            "kind": "IntValue",
            "value": "5"
          }
        }],
        "directives": [],
        "selectionSet": {
          "kind": "SelectionSet",
          "selections": [{
            "kind": "FragmentSpread",
            "name": {
              "kind": "Name",
              "value": "UserEntry4"
            },
            "directives": []
          }]
        }
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 371,
    "source": {
      "body": "fragment UserEntry1 on User {\\n  firstName\\n}\\n\\nfragment UserEntry3 on User {\\n  id\\n}\\n#import \\"./fragment3.graphql\\"\\n\\nfragment UserEntry5 on User {\\n  ...UserEntry4\\n}\\n#import \\"./fragment1.graphql\\"\\n#import \\"./fragment4.graphql\\"\\n\\nfragment UserEntry4 on User {\\n  ...UserEntry1\\n  ...UserEntry5\\n}\\n#import './fragment3.graphql'\\n\\nquery User {\\n  user(id: 5) {\\n    ...UserEntry4\\n  }\\n}\\n\\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};

`;

exports[`macros [loader] without fragment: [loader] without fragment 1`] = `

import { loader } from 'graphql.macro';
const query = loader('./fixtures/query.graphql');

      ↓ ↓ ↓ ↓ ↓ ↓

const query = {
  "kind": "Document",
  "definitions": [{
    "kind": "OperationDefinition",
    "operation": "query",
    "name": {
      "kind": "Name",
      "value": "User"
    },
    "variableDefinitions": [],
    "directives": [],
    "selectionSet": {
      "kind": "SelectionSet",
      "selections": [{
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "user"
        },
        "arguments": [{
          "kind": "Argument",
          "name": {
            "kind": "Name",
            "value": "id"
          },
          "value": {
            "kind": "IntValue",
            "value": "5"
          }
        }],
        "directives": [],
        "selectionSet": {
          "kind": "SelectionSet",
          "selections": [{
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "firstName"
            },
            "arguments": [],
            "directives": []
          }, {
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "lastName"
            },
            "arguments": [],
            "directives": []
          }]
        }
      }]
    }
  }],
  "loc": {
    "start": 0,
    "end": 62,
    "source": {
      "body": "query User {\\n  user(id: 5) {\\n    firstName\\n    lastName\\n  }\\n}\\n",
      "name": "GraphQL request",
      "locationOffset": {
        "line": 1,
        "column": 1
      }
    }
  }
};

`;
